Release GIL before destructing ProcessGroup classes

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56381 Release GIL before destructing ProcessGroup classes**

Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

ezyang committed:
Release GIL before destructing ProcessGroup classes

Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]

ezyang commented:
This patch is known not to work because apparently I haven't nailed down the finalizing logic correctly

ezyang committed:
Update on "Release GIL before destructing ProcessGroup classes"


Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]

ezyang commented:
Whelp turns out release gil is not safe without an enclosing GIL. New patch should work.

ezyang commented:
I can't easily tell if this is a preexisting condition or an exciting new failure:

```
Apr 19 17:26:17   test_DistributedDataParallel_non_default_stream (__main__.TestDistBackendWithSpawn) ... terminate called after throwing an instance of 'c10::Error'
Apr 19 17:26:17   what():  CUDA error: driver shutting down
Apr 19 17:26:17 Exception raised from query at /var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDAEvent.h:95 (most recent call first):
Apr 19 17:26:17 frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f2d00c0c8db in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
Apr 19 17:26:17 frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xce (0x7f2d00bfbc5e in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
Apr 19 17:26:17 frame #2: c10d::ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const + 0x104 (0x7f2d1c2d9c14 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
Apr 19 17:26:17 frame #3: c10d::ProcessGroupNCCL::workCleanupLoop() + 0x288 (0x7f2d1c2dc6d8 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
Apr 19 17:26:17 frame #4: <unknown function> + 0xc819d (0x7f2d1f69919d in /opt/conda/lib/libstdc++.so.6)
Apr 19 17:26:17 frame #5: <unknown function> + 0x76ba (0x7f2d55ad26ba in /lib/x86_64-linux-gnu/libpthread.so.0)
Apr 19 17:26:17 frame #6: clone + 0x6d (0x7f2d558084dd in /lib/x86_64-linux-gnu/libc.so.6)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36779: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] THPPointer<_object>::free()+0xc(0x7f2d1bee56cc) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [3] ???+0x1bc584ac(0x7f2d1bc584ac) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [4] ???+0x5573b008(0x7f2d5573b008) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [5] ???+0x5573b055(0x7f2d5573b055) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [6] ???(0x559091ac52d9) in /opt/conda/bin/python
Apr 19 17:26:17 [7] ???(0x559091ac53b8) in /opt/conda/bin/python
Apr 19 17:26:17 [8] PyErr_PrintEx+0x32(0x559091ac5422) in /opt/conda/bin/python
Apr 19 17:26:17 [9] PyRun_SimpleStringFlags+0x66(0x559091acb706) in /opt/conda/bin/python
Apr 19 17:26:17 [10] Py_Main+0x423(0x559091acf4e3) in /opt/conda/bin/python
Apr 19 17:26:17 [11] main+0xee(0x559091999b0e) in /opt/conda/bin/python
Apr 19 17:26:17 [12] __libc_start_main+0xf0(0x7f2d55721840) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [13] ???(0x559091a78d70) in /opt/conda/bin/python
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36786: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] accept4+0x68(0x7f2d55809988) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [3] ???(0x7f2cddbfbce3) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [4] ???(0x7f2cddc9dc36) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [5] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [6] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [7] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36787: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] pthread_cond_timedwait+0x129(0x7f2d55ad8709) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [3] c10d::ProcessGroupNCCL::ncclCommWatchdogInternal()+0xad3(0x7f2d1c2df153) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [4] c10d::ProcessGroupNCCL::ncclCommWatchdog()+0x81(0x7f2d1c2e05b1) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [5] ???+0x1f69919d(0x7f2d1f69919d) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [6] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [7] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [8] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36788: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x11f29127(0x7f2d11f29127) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [2] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [3] gsignal+0x38(0x7f2d55736438) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [4] abort+0x16a(0x7f2d5573803a) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [5] __gnu_cxx::__verbose_terminate_handler()+0xbc(0x7f2d1f67e84a) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [6] ???+0x1f67cf47(0x7f2d1f67cf47) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [7] ???+0x1f67cf7d(0x7f2d1f67cf7d) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [8] __cxa_rethrow+0(0x7f2d1f67d15a) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [9] c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xfa(0x7f2d00bfbc8a) in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so
Apr 19 17:26:17 [10] c10d::ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const+0x104(0x7f2d1c2d9c14) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [11] c10d::ProcessGroupNCCL::workCleanupLoop()+0x288(0x7f2d1c2dc6d8) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [12] ???+0x1f69919d(0x7f2d1f69919d) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [13] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [14] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [15] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36792: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] __poll+0x2d(0x7f2d557fc80d) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [3] ???(0x7f2cddca7bc1) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [4] ???(0x7f2cddc80eea) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [5] ???(0x7f2cddc9dc36) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [6] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [7] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [8] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36797: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] pthread_cond_wait+0xc0(0x7f2d55ad8360) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [3] ???+0x444dbc8(0x7f2d0444dbc8) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so
Apr 19 17:26:17 [4] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [5] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [6] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36799: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] pthread_cond_timedwait+0x129(0x7f2d55ad8709) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [3] PyEval_RestoreThread+0x2b8(0x5590919bc3e8) in /opt/conda/bin/python
Apr 19 17:26:17 [4] torch::autograd::python::PythonEngine::thread_init(int, std::shared_ptr<torch::autograd::ReadyQueue> const&, bool)+0xd5(0x7f2d1bc59a75) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [5] ???+0x1f69919d(0x7f2d1f69919d) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [6] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [7] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [8] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36800: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] pthread_cond_timedwait+0x129(0x7f2d55ad8709) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [3] PyEval_RestoreThread+0x2b8(0x5590919bc3e8) in /opt/conda/bin/python
Apr 19 17:26:17 [4] torch::autograd::python::PythonEngine::thread_init(int, std::shared_ptr<torch::autograd::ReadyQueue> const&, bool)+0xd5(0x7f2d1bc59a75) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so
Apr 19 17:26:17 [5] ???+0x1f69919d(0x7f2d1f69919d) in /opt/conda/lib/libstdc++.so.6
Apr 19 17:26:17 [6] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [7] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [8] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 SIGABRT(6), PID: 36779, Thread 36803: 
Apr 19 17:26:17 [0] ???+0x11f28312(0x7f2d11f28312) in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so
Apr 19 17:26:17 [1] ???+0x55adc390(0x7f2d55adc390) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [2] __poll+0x2d(0x7f2d557fc80d) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [3] ???(0x7f2cddca7bc1) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [4] ???(0x7f2cddc80eea) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [5] ???(0x7f2cddc9dc36) in /usr/lib/x86_64-linux-gnu/libcuda.so.1
Apr 19 17:26:17 [6] ???+0x55ad26ba(0x7f2d55ad26ba) in /lib/x86_64-linux-gnu/libpthread.so.0
Apr 19 17:26:17 [7] clone+0x6d(0x7f2d558084dd) in /lib/x86_64-linux-gnu/libc.so.6
Apr 19 17:26:17 [8] ???(0)
Apr 19 17:26:17 
Apr 19 17:26:17 FAIL (5.231s)
```

ezyang committed:
Update on "Release GIL before destructing ProcessGroup classes"


Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]

ezyang committed:
Update on "Release GIL before destructing ProcessGroup classes"


Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]

ezyang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -38,6 +38,42 @@
+  // This ctor is very important; see
+  // https://github.com/pybind/pybind11/issues/2957
+  IntrusivePtrNoGilDestructor(T* impl) : impl_(c10::intrusive_ptr<T>::reclaim(impl)) {
+    c10::raw::intrusive_ptr::incref(impl);
```
@wanchaol I'm tagging you specifically for this line, this replicates the logic you wrote in https://github.com/pytorch/pytorch/pull/47586/files#diff-a8f056e4406fefdc132b5ba55fcebb64757224f649a1acd86b1861fa4fb3a5bcR246 because your implementation is not public. Should we be adding another static method for this specific use case? Note that I *cannot* use `unsafe_reclaim_from_nonowning` as this method requires refcount to be nonzero, which is not the case here (the raw pointer constructor is being used in the classic unique pointer manner)

wanchaol replied:
Yeah it looks reasonable to have a general static function to be used by both. For this line specifically, it seems not increasing the weak_count, if making this a general function, i guess weak_count also need to be increased?

ezyang replied:
incref increases weakcount, not to worry :)

ezyang committed:
Update on "Release GIL before destructing ProcessGroup classes"


Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]

ezyang committed:
Update on "Release GIL before destructing ProcessGroup classes"


Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -38,6 +38,49 @@
+    c10::raw::intrusive_ptr::incref(impl);
+  }
+  ~IntrusivePtrNoGilDestructor() {
+    if (impl_) {
```
is it because we need to keep intrusive_ptr python-free (couldn't put this logic directly in intrusive_ptr), that we have to define a separate class specifically for the destructor on GIL?

ezyang replied:
not just that, but if the destructor doesn't actually block and is quick to run, releasing the GIL does more harm than good (it costs performance)

wanchaol reviewed (approved):
looks good to me, only have some questions.

ezyang committed:
Update on "Release GIL before destructing ProcessGroup classes"


Part of fix for https://github.com/pytorch/pytorch/issues/56297

Differential Revision: [D27855337](https://our.internmc.facebook.com/intern/diff/D27855337/)

[ghstack-poisoned]
